compiler/rustc_next_trait_solver/src/solve/trait_goals.rs

```diff
@@ -1147,13 +1148,86 @@ where
             ecx.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
         })
     }
+}
+
+/// How we've proven this trait goal.
```

Comment:
pls document how this is used, particularly bc the choice of what variant you return affects behavior

Reply:
also, woild it not be too verbose to call this something like TraitBoundProvenVia? Then it's obvious in the non-trait bound merging function that we use this based on how we proved the corresponding trait bound.